File: tools/math_dsl/ways_evaluator.py
"""W244 wave 25 — closed-form analytical model for `ways_evaluator`.

Industry pattern (BTG Megaways: Bonanza, Extra Chilli, White Rabbit;
Microgaming 243-ways, 1024-ways; NetEnt Twin Spin 243-ways;
Pragmatic Big Bass Splash 4096-ways):

  Variable-rows / fixed-ways game
  -------------------------------
    Slot has N_REELS reels. Each reel reveals between `min_rows` and
    `max_rows` symbols on a given spin (random for Megaways, fixed
    for traditional 243/1024-ways).

    Ways count per spin = product over reels(row_count_per_reel).

    Per-way pay: standard left-to-right matching with symbol weights
    + paytable. Operator supplies the per-way RTP for the configured
    paytable + reel composition.

  Closed-form RTP contribution
  ----------------------------
    Per-spin RTP = E[ways_count] × per_way_rtp_x_bet

    where E[ways_count] = product over reels(E[row_count_per_reel])
    when reels are independent.

    For variable-rows topology (Megaways), each reel's row count
    follows an operator-supplied empirical distribution:
      `row_distribution[reel_idx]: {row_count: probability}`

    Closed-form scales O(n_reels × max_rows).

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_ways_evaluator_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_ways_evaluator_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class WaysEvaluatorParams:
    """Closed-form model inputs."""
    # Per-reel row count distribution. Element i is a dict
    # {row_count: probability} summing to 1.0 per reel.
    row_distribution_per_reel: tuple[dict[int, float], ...]
    # Per-way RTP — what each "way" contributes on average per spin.
    # Industry typical: ~ 0.96 / E[ways] so total = ~96 % when × E[ways].
    per_way_rtp_x_bet: float

    def __post_init__(self):
        if not self.row_distribution_per_reel:
            raise ValueError("row_distribution_per_reel must be non-empty")
        if self.per_way_rtp_x_bet < 0:
            raise ValueError("per_way_rtp_x_bet must be ≥ 0")
        for i, dist in enumerate(self.row_distribution_per_reel):
            if not isinstance(dist, dict) or not dist:
                raise ValueError(
                    f"reel {i} row_distribution must be non-empty dict"
                )
            for rows, prob in dist.items():
                if rows < 1:
                    raise ValueError(
                        f"reel {i}: row count {rows} must be ≥ 1"
                    )
                if prob < 0:
                    raise ValueError(
                        f"reel {i}: probability {prob} must be ≥ 0"
                    )
            s = sum(dist.values())
            if abs(s - 1.0) > 1e-9:
                raise ValueError(
                    f"reel {i}: probabilities sum to {s}, expected 1.0"
                )


def expected_rows_per_reel(
    params: WaysEvaluatorParams,
) -> tuple[float, ...]:
    """E[row_count] per reel."""
    return tuple(
        sum(rows * prob for rows, prob in dist.items())
        for dist in params.row_distribution_per_reel
    )


def expected_ways_count(params: WaysEvaluatorParams) -> float:
    """E[ways] = product over reels(E[row_count_per_reel]).

    Valid only if reels are independent (industry-standard assumption).
    """
    e_rows = expected_rows_per_reel(params)
    out = 1.0
    for e in e_rows:
        out *= e
    return out


def ways_evaluator_rtp(params: WaysEvaluatorParams) -> dict:
    """Per-spin RTP + per-reel breakdown."""
    e_rows = expected_rows_per_reel(params)
    e_ways = expected_ways_count(params)
    rtp = e_ways * params.per_way_rtp_x_bet

    per_reel = []
    for i, (dist, e_r) in enumerate(zip(params.row_distribution_per_reel, e_rows)):
        per_reel.append({
            "reel_index": i,
            "expected_rows": e_r,
            "row_distribution": dict(dist),
        })

    return {
        "rtp_contribution": rtp,
        "n_reels": len(params.row_distribution_per_reel),
        "expected_rows_per_reel": list(e_rows),
        "expected_ways_count": e_ways,
        "per_way_rtp_x_bet": params.per_way_rtp_x_bet,
        "per_reel_breakdown": per_reel,
    }

File: rust-sim/src/kernels/ways_evaluator.rs
//! W244.25 — ways_evaluator Rust port.
//!
//! Closed-form: E[ways] = product(E[row_count_per_reel]) under reel independence.
//! RTP = E[ways] × per_way_rtp_x_bet.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WaysEvaluatorParams {
    /// Per-reel row count distribution. Element i is {row_count: probability}.
    pub row_distribution_per_reel: Vec<BTreeMap<u32, f64>>,
    pub per_way_rtp_x_bet: f64,
}

impl WaysEvaluatorParams {
    pub fn validate(&self) -> Result<(), String> {
        if self.row_distribution_per_reel.is_empty() {
            return Err("row_distribution_per_reel must be non-empty".into());
        }
        if self.per_way_rtp_x_bet < 0.0 {
            return Err("per_way_rtp_x_bet must be ≥ 0".into());
        }
        for (i, dist) in self.row_distribution_per_reel.iter().enumerate() {
            if dist.is_empty() {
                return Err(format!("reel {} distribution must be non-empty", i));
            }
            for (&rows, &prob) in dist {
                if rows < 1 {
                    return Err(format!("reel {}: row count {} must be ≥ 1", i, rows));
                }
                if prob < 0.0 {
                    return Err(format!("reel {}: probability {} must be ≥ 0", i, prob));
                }
            }
            let s: f64 = dist.values().sum();
            if (s - 1.0).abs() > 1e-9 {
                return Err(format!(
                    "reel {}: probabilities sum to {}, expected 1.0", i, s
                ));
            }
        }
        Ok(())
    }
}

pub fn expected_rows_per_reel(p: &WaysEvaluatorParams) -> Vec<f64> {
    p.row_distribution_per_reel.iter()
        .map(|dist| dist.iter().map(|(&r, &prob)| (r as f64) * prob).sum())
        .collect()
}

pub fn expected_ways_count(p: &WaysEvaluatorParams) -> f64 {
    expected_rows_per_reel(p).iter().product()
}

#[derive(Debug, Serialize)]
pub struct PerReel {
    pub reel_index: u32,
    pub expected_rows: f64,
    pub row_distribution: BTreeMap<u32, f64>,
}

#[derive(Debug, Serialize)]
pub struct WaysEvaluatorResult {
    pub rtp_contribution: f64,
    pub n_reels: u32,
    pub expected_rows_per_reel: Vec<f64>,
    pub expected_ways_count: f64,
    pub per_way_rtp_x_bet: f64,
    pub per_reel_breakdown: Vec<PerReel>,
}

pub fn ways_evaluator_rtp(p: &WaysEvaluatorParams) -> WaysEvaluatorResult {
    let e_rows = expected_rows_per_reel(p);
    let e_ways: f64 = e_rows.iter().product();
    let rtp = e_ways * p.per_way_rtp_x_bet;

    let per_reel: Vec<PerReel> = p.row_distribution_per_reel.iter()
        .enumerate()
        .zip(e_rows.iter())
        .map(|((i, dist), &e_r)| PerReel {
            reel_index: i as u32,
            expected_rows: e_r,
            row_distribution: dist.clone(),
        })
        .collect();

    WaysEvaluatorResult {
        rtp_contribution: rtp,
        n_reels: p.row_distribution_per_reel.len() as u32,
        expected_rows_per_reel: e_rows,
        expected_ways_count: e_ways,
        per_way_rtp_x_bet: p.per_way_rtp_x_bet,
        per_reel_breakdown: per_reel,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fixed_243_ways() {
        let mut dist = BTreeMap::new();
        dist.insert(3, 1.0);
        let p = WaysEvaluatorParams {
            row_distribution_per_reel: vec![dist.clone(); 5],
            per_way_rtp_x_bet: 0.96 / 243.0,
        };
        let r = ways_evaluator_rtp(&p);
        assert!((r.expected_ways_count - 243.0).abs() < 1e-10);
        assert!((r.rtp_contribution - 0.96).abs() < 1e-10);
    }

    #[test]
    fn megaways_117649() {
        let mut dist = BTreeMap::new();
        dist.insert(7, 1.0);
        let p = WaysEvaluatorParams {
            row_distribution_per_reel: vec![dist; 6],
            per_way_rtp_x_bet: 0.96 / 117649.0,
        };
        let r = ways_evaluator_rtp(&p);
        assert!((r.expected_ways_count - 117649.0).abs() < 1e-6);
    }
}
